Fill.py

```diff
@@ -1086,3 +1120,76 @@ def failed(warning: str, force: typing.Union[bool, str]) -> None:
         except Exception as e:
             raise Exception(
                 f"Error running plando for player {player} ({multiworld.player_name[player]})") from e
+
+
+def distribute_local_nonprogression(multiworld: MultiWorld,
+                                    fill_locations: typing.List[Location],
+                                    useful_itempool: typing.List[Item],
+                                    filler_itempool: typing.List[Item]) \
+        -> typing.Tuple[typing.List[Item], typing.List[Item], typing.List[Location],
+                        typing.List[Location], typing.List[Location]]:
+    # call remaining_fill early on local items, but sort the list of locations filled this way to the back of the list
+    # instead of popping them entirely
+    # sort the location list
+    locations: typing.Dict[LocationProgressType, typing.List[Location]] = {
+        loc_type: [] for loc_type in LocationProgressType}
+
+    for loc in fill_locations:
+        locations[loc.progress_type].append(loc)
+
+    priority_locations = locations[LocationProgressType.PRIORITY]
+    default_locations = locations[LocationProgressType.DEFAULT]
+    excluded_locations = locations[LocationProgressType.EXCLUDED]
+    # recombine
+    fill_locations = [*excluded_locations, *default_locations, *priority_locations]
+
+    local_nonprogression_items: typing.List[Item] = []
+    # filter out local filler
+    new_filler_itempool: typing.List[Item] = []
+    for item in filler_itempool:
+        if item.name in multiworld.worlds[item.player].options.local_items.value:
```

Comment:
I'm not sure this accounts for item link local items. Though I'm not currently sure if it can/should account for them. Item links are confusing.

Item links can define their own local items and non-local items, the rules for which are set alongside normal local items/non-local items rules in `worlds.generic.Rules.locality_rules()`.

Reply:
`filler_itempool` can be large (`useful_itempool` typically less so). I would prefer to create a more direct dictionary, of local items sets per player, outside the loops and then access that from within the loops, reducing the number of attribute lookups+accesses within the loops.

Before the `for item in filler_itempool:` loop:
`local_items_by_player = {world.player: world.options.local_items.value for world in multiworld.worlds}`
Inside the loop:
`    if item.name in local_items_by_player[item.player]:`